deploy/StableJoeStaking.js

```diff
@@ -12,33 +17,45 @@ module.exports = async function ({ getNamedAccounts, deployments }) {
     rewardToken = "0x9Ad6C38BE94206cA50bb0d90783181662f0Cfa10";
     feeCollector = "0x2fbB61a10B96254900C03F1644E9e1d2f5E76DD2";
     depositFeePercent = 0;
-    proxyOwner = deployer.address;
+    smolJoes = "0xce347E069B68C53A9ED5e7DA5952529cAF8ACCd4"; // use an ERC20 as placeholder for now
+    proxyOwner = deployer;
   } else if (chainId == 43114 || chainId == 31337) {
     // avalanche mainnet or hardhat network addresses
     joeAddress = "0x6e84a6216eA6dACC71eE8E6b0a5B7322EEbC0fDd";
     // USDC.e
     rewardToken = "0xB97EF9Ef8734C71904D8002F8b6Bc66Dd9c48a6E";
     feeCollector = "0x2fbB61a10B96254900C03F1644E9e1d2f5E76DD2";
     depositFeePercent = 0;
+    smolJoes = "0xC70DF87e1d98f6A531c8E324C9BCEC6FC82B5E8d";
     proxyOwner = "0x2fbB61a10B96254900C03F1644E9e1d2f5E76DD2";
   }
 
-  await catchUnknownSigner(
-    deploy("StableJoeStaking", {
+  await catchUnknownSigner(async () => {
+    const sJoe = await deploy("StableJoeStaking", {
       from: deployer,
       proxy: {
         owner: proxyOwner,
         proxyContract: "OpenZeppelinTransparentProxy",
         execute: {
           init: {
-            methodName: "initialize",
-            args: [rewardToken, joeAddress, feeCollector, depositFeePercent],
+            methodName: "initializeSmolJoes",
```

Comment:
You shouldn't change this block (`init`) as it won't get called on upgrade anyway (plus it could have unintended consequences for the unit tests if they use these deployment scripts for fixture generation). If you want to use hardhat-deploy to set the address you can add an `onUpgrade` block

```
    init: {...unchanged...},
    onUpgrade: {
        methodName: "initializeSmolJoes",
        args: [smolJoes],
    }
```
To do this however you will need to make the `initializeSmolJoes` function idempotent (not reverting if `smolJoes` variable is non 0) as it will be called every time the function gets upgraded.

Reply:
That's very cool, I didn't know. Thanks for sharing that!
However, if we remove that check, then anyone can set it to any address. The new `Initializable` (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/utils/Initializable.sol) from OZ makes a lot of sense in that case.

Reply:
I tried to use `onUpgrade` but it would revert without any reason when I ran the upgrade locally. On the contrary, when I didn't use `onUpgrade` it would upgrade fine but obviously I had to call `initializeSmolJoes` manually:

Hey, I have a question about proxy upgrades. I have a proxy contract `Test` that I want to upgrade. On the upgrade, I want it to call a function `intitializeAgain()`.

In my deploy script I have:

```
    await deploy("Test", {
      from: deployer,
      proxy: {
        owner: deployer,
        proxyContract: "OpenZeppelinTransparentProxy",
        execute: {
          init: {
            methodName: "initialize",
            args: [
              param1,
              param2,
            ],
          },
          onUpgrade: {
            methodName: "initializeAgain",
            args: [param3],
          },
        },
      },
      log: true,
    });
```

When I run this locally and try to upgrade the implementation after making a modification to the `Test` contract, it reverts without any reason:

```
 Compilation finished successfully
Creating Typechain artifacts in directory types for target ethers-v5
Successfully generated Typechain artifacts!
deployer is:  0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
proxyOwner is:  0x2fbB61a10B96254900C03F1644E9e1d2f5E76DD2
reusing "DefaultProxyAdmin" at 0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0
deploying "StableJoeStaking_Implementation" (tx: 0x44ad6abb7bc903b949f19176000973c787c83e1b16f92ebbca42e6a9cec86934)...: deployed at 0x0165878A594ca255338adfa4d48449f69242Eb8F with 3619599 gas
An unexpected error occurred:

Error: ERROR processing /Users/cryptofish/workspace/joe-core/deploy/StableJoeStaking.js:
ProviderError: Error: Transaction reverted without a reason string
```

The deploy code looks like this:

```
    const sJoe = await deploy("StableJoeStaking", {
      from: deployer,
      proxy: {
        owner: deployer, // change from proxyOwner so there are no ownership issues on upgrading
        proxyContract: "OpenZeppelinTransparentProxy",
        execute: {
          init: {
            methodName: "initialize",
            args: [
              rewardToken,
              joeAddress,
              feeCollector,
              depositFeePercent,
              smolJoes,
            ],
          },
          onUpgrade: {
            methodName: "initializeSmolJoes",
            args: [smolJoes],
          },
        },
      },
      log: true,
    });

```

Reply:
So it looks like this is failing because of the `onlyOwner` modifier on `initializeSmolJoes`. This is because hardhat-deploy will use `upgradeToAndCall` function on the `ProxyAdmin` contract which performs the upgrade and then calls `initializeSmolJoes`. As the `ProxyAdmin` address isn't the owner of the `StableJoeStaking` contract this fails. I confirmed this by calling `transferOwnership` with the `ProxyAdmin` address on `StableJoeStaking` and then performing the upgrade, in this case the `initializeSmolJoes` function didn't revert as `msg.sender` was the owner (The proxy admin address). This makes sense as the `upgradeToAndCall` function should be atomic (the upgrade and function call should succeed or the tx should rollback). 

AFAICT there is two options here:
1. Change the ownership of `StableJoeStaking` to the `ProxyAdmin` contract.
2. Not use atomic upgrade and call the `initializeSmolJoes` function after the contract has been upgraded, this option is probably fine as the smolJoes address **isn't** needed for the contract to function.

Reply:
Thanks for spotting that. Regarding the two options - option 1 is a pain since several other functions are `onlyOwner` so you'll need to transfer ownership of `StableJoeStaking` back to the multisig. 

Can't put `initializeSmolJoes` under the `initializer` modifier either. If only we could use `reinitializer` in the latest OZ `Initializable`, but it's v8 so not compatible. 

Anyways, gone with option 2 here and turned `initializeSmolJoes` to `setSmolJoes`.

Reply:
Option 1 would be a big mistake: Transparent Proxies don't allow the owner of the proxy to call any function of the implementation. So if the ownership is transferred, you couldn't call any `onlyOwner` function, nor transfer the ownership again.